File: Challenges/Problem_17/Number_letter_counts.js
/// KuntaNgcobo
// Number Letter Counts

let units = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", 
	"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]

let tens = ["", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]

let hundred = "hundred"

let thouOrder = ["", "thousand", "million", "billion", "trillion", "quadrillion"]

let reverse = str => {return [...str].reverse().join('')}

let word_number = number => {
	if(number >= 10**18){
    return "Number Too Large"
  }

	let strNumber = reverse(number + "")
	let length = strNumber.length
	let thirdOrder = []

	for(var i = 0; i < length; i += 3){
		thirdOrder = thirdOrder.concat( parseInt( reverse(strNumber.substring(i, i + 3) ) ) )
  }

	let word = ""
	for(var i = 0; i < thirdOrder.length; i++){
		let temp = ""
		let num = thirdOrder[i]

		let ten = num%100
		let hundreds = parseInt(num/100)
		
    if(ten < 20){
			temp += units[ten]
    } else {
			let dunit = parseInt(ten/10)
			let unit = ten%10
			temp += tens[dunit - 1] + "" + units[unit]
    }

		if(hundreds > 0){
			temp = units[hundreds] + "" + hundred + ( ten > 0 ? "and" : "" ) + temp
    }

		word = temp + "" + thouOrder[i] + "" + word
	}

	return word
}

let numberLetterCounts = limit => {
	let length = 0
	for(var i = 0; i < limit + 1; i++){
		length += word_number(i).length
	}

	return length
}

console.log(5, numberLetterCounts(5))

console.log(150, numberLetterCounts(150))

console.log(1000, numberLetterCounts(1000))

File: Challenges/Problem_17/Number_letter_counts.py
# KuntaNgcobo
# Number Letter Counts

units = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", 
	"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]

tens = ["", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]

hundred = "hundred"

thouOrder = ["", "thousand", "million", "billion", "trillion", "quadrillion"]

def word_number(number):
	if(number >= 10**18):
		return "Number Too Large"
	strNumber = str(number)[::-1]
	length = len(strNumber)
	thirdOrder = []


	for i in range(0, length, 3):
		thirdOrder.append( int( strNumber[i : i + 3][::-1] ) )

	word = ""
	for i in range( len(thirdOrder) ):
		temp = ""
		num = thirdOrder[i]

		ten = num%100
		hundreds = num//100
		if(ten < 20):
			temp += units[ten]
		else:
			dunit = ten//10
			unit = ten%10
			temp += tens[dunit - 1] + "" + units[unit]

		if(hundreds > 0):
			temp = units[hundreds] + "" + hundred + ( "and" if(ten > 0) else "" ) + temp

		word = temp + "" + thouOrder[i] + "" + word

	return word

def numberLetterCounts(limit):
	length = 0
	for i in range(1, limit + 1):
		length += len( word_number(i) )

	return length

def printer(number, wordShow = True):
	if(word_number):
		print(number, word_number( number ), len(word_number( number )))
	else:
		print(number, len(word_number( number )))

print(5, numberLetterCounts(5))

print(150, numberLetterCounts(150))

print(1000, numberLetterCounts(1000))